### ISSUE TITLE
Potential use after free

### ISSUE THREAD

**Issue by snnn** [microsoft/onnxruntime#11392]
**Describe the bug**

Assume your have an exe, which has code like:

```c++
Environment& GetEnv() {
  static std::unique_ptr<Environment> session_env = InitializeEnv();
  return *session_env;
}
```
Then the destructor of session_env  will be called when the process is shutting down, which is too late if you have dynamically loaded any EP. Because, when you load an EP, if the EP has any global variables with non-trivial destructors, they will register themselves  to C/C++ runtime's atexit function. Because they are registered later than the onnxruntime::Environment object, they get destroyed earlier too. So, it would be like:

1. construct onnxruntime::Environment. And register the object to the C/C++ runtime's atexit function
2. Load CUDA EP and init its global variables. And register them to the C/C++ runtime's atexit function
3. Destroy CUDA EP's global variables
4. Start to destroy onnxruntime::Environment
5. call the shutdown method of the CUDA EP (here it should not read any global variables, but if it doesn't, it doesn't need to exist at first
6. Finish destroying onnxruntime::Environment

Basically, when onnxruntime::Environment::~Environment() is called, it is too late if we already started to destroy global variables.  But for our python implementation, I don't see a way to solve this. It would be either too late or too early.

**Comment by snnn** [comment]
I couldn't find a solution for it.

**Comment by yufenglee** [comment]
why not do step 5 before 3?

**Comment by snnn** [comment]
Step 4 calls step 5.
